test/e2e/storage/csi_mock_volume.go

```diff
@@ -889,12 +889,14 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 				ginkgo.By("Checking PVC events")
 				nodeAnnotationSet := false
 				nodeAnnotationReset := false
+				watchFailed := false
 			loop:
 				for {
 					select {
 					case event, ok := <-pvcWatch.ResultChan():
 						if !ok {
-							framework.Failf("PVC watch ended prematurely")
```

Comment:
the main issue seems to be opening a raw watch and expecting it to never drop during the test. shouldn't we be using the watch tools helper that cam reestsblish a dropped watch and continue it?

Reply:
cc @tnozicka

Reply:
We've discussed that in more detail in https://github.com/kubernetes/kubernetes/pull/93658#pullrequestreview-460530486: even these helpers only guarantee that eventually the current state is reported.

What the test would like to have (but cannot get reliably) is a list of all intermediate states. If it isn't able to get that, it cannot make assertions about these states, but that's not a test failure - at least that's how I have implemented it in this PR.

Reply:
IMHO it is better to have a test that most of the time is able to check something than to have no test for that at all.

Reply:
> even these helpers only guarantee that eventually the current state is reported.

with watch bookmarks, they are far more reliable than a raw watch…

Reply:
That sounds worthwhile. Where can I find these watch tool helpers?

Reply:
NewIndexerInformerWatcher (uses an informer, falls back to list if watch cannot be reestablished) or NewRetryWatcher (retries watch only) in k8s.io/client-go/tools/watch

Reply:
After reading the documentation, it is not clear whether NewIndexerInformerWatcher is really less likely to miss intermediate object updates. NewRetryWatcher sounds more promising, but also is "not resilient to etcd cache not having the resource version anymore". Not sure how it then behaves.

Both seem have in common that they hide from the user when changes have been lost, right?

From the perspective of this test, that's worse than loosing the raw watch. When the raw watch fails, the test knows that it doesn't have all data. But when the watch tools internally hide such a failure, the test might flake because it makes assumption about incomplete data. I'm leaning towards keeping the test as it is now.

If there are concerns that it will not test the feature because the watch fails too frequently, perhaps adding a retry loop around the entire test with a limit and "test passed if limit reached" would address that?

Reply:
> NewRetryWatcher sounds more promising, but also is "not resilient to etcd cache not having the resource version anymore".

With watch bookmarks enabled, we expect that not to happen under normal circumstances

Reply:
> perhaps adding a retry loop around the entire test with a limit and "test passed if limit reached" would address that?

I would prefer deflaking the test over simply retrying it, especially since the storage tests already consume a disproportionate amount of time in the e2e

Reply:
> With watch bookmarks enabled, we expect that not to happen under normal circumstances

In that case I'll switch to that.

/hold

Reply:
Done.

/cancel hold

Reply:
/hold cancel